### PR TITLE
bpo-1635741: Port _gdbm module to multiphase initialization

### PR DESCRIPTION
**Pull request by corona10**
https://bugs.python.org/issue1635741

### CODE REVIEW

**Comment by corona10** [comment]
I' ve checked the memory leak with these scripts and no leaks found

```
./python.exe -m test test_dbm_gnu -R 3:3
```

```  python
def test_dbm(self):
    code = textwrap.dedent(r"""
        import glob

        import test.support
        dbm = test.support.import_module('_gdbm')
        # or dbm = test.support.import_module('dbm')
        _fname = test.support.TESTFN

        def delete_files():
            for f in glob.glob(_fname + "*"):
                test.support.unlink(f)

        f = dbm.open(_fname, 'n')
        f[b'g'] = b"indented"
        f.close()
       delete_files()
    """)
    ret = test.support.run_in_subinterp(code)
    self.assertEqual(ret, 0)
```

**Comment by corona10** [comment]
Note that `dbm = test.support.import_module('dbm')` is leaked at the master branch :)

File: Modules/_gdbmmodule.c [hunk 1]
```diff
@@ -529,40 +546,31 @@ static PyMethodDef dbm_methods[] = {
...
+    {0, 0}
+};
+
+static PyType_Spec gdbmtype_spec = {
+    .name = "_gdbm.gdbm",
+    .basicsize = sizeof(gdbmobject),
+    // Calling PyType_GetModuleState() on a subclass is not safe.
+    // dbmtype_spec does not have Py_TPFLAGS_BASETYPE flag
+    // which prevents to create a subclass.
+    // So calling PyType_GetModuleState() in this file is always safe.
```
Comment (vstinner) [inline Modules/_gdbmmodule.c:572]: I confirm that it's not possible to create a subclass:

```
>>> import _gdbm
>>> db=_gdbm.open("test", "c")
>>> class MyDB(type(db)): pass
... 
TypeError: type '_gdbm.gdbm' is not an acceptable base type
```

**Review by vstinner — APPROVED** [review]
LGTM.

Minor remarks:
* I'm not sure why only a few methods uses the defining class, whereas some others use Py_TYPE(self).
* nitpick: some "clinic input" uses an empty line betwen the function name and its parameters, sometimes there is none.

But I don't think that it's worth it to address these minor things. *If* you care, you may also update _dbmmodule.c. Maybe in a follow-up PRs.

**Comment by vstinner** [comment]
> Note that dbm = test.support.import_module('dbm') is leaked at the master branch :)

Maybe open an issue to track this bug.

**Comment by vstinner** [comment]
> Maybe open an issue to track this bug.

Well, I added a comment to: https://bugs.python.org/issue40987#msg371679

**Comment by corona10** [comment]
@vstinner 
> Maybe open an issue to track this bug

Oh I mean the subinterpreter test with `dbm = test.support.import_module('dbm')`is leaked
With this change, there is no leak.

**Comment by corona10** [comment]
@vstinner Thanks for the review!

**Comment by vstinner** [comment]
>  With this change, there is no leak.

Oh ok, that's great! Sorry, I misunderstood your comment.

**Pull request merged**